txpool/txpool.go

```diff
@@ -288,89 +302,128 @@ func (t *TxPool) validateTx(tx *types.Transaction) error {
 	return nil
 }
 
-type txQueue struct {
-	txs       txHeap
+// txHeapWrapper is a wrapper object for account based transactions
+type txHeapWrapper struct {
+	// txs is the actual min heap (nonce ordered) for account transactions
+	txs txHeap
+
+	// nextNonce is a field indicating what should be the next
+	// valid nonce for the account transaction
 	nextNonce uint64
 }
 
-func newTxQueue() *txQueue {
-	return &txQueue{
+// newTxHeapWrapper creates a new account based tx heap
+func newTxHeapWrapper() *txHeapWrapper {
+	return &txHeapWrapper{
 		txs: txHeap{},
 	}
 }
 
-func (t *txQueue) Reset(txns ...*types.Transaction) {
-	lowestNonce := t.nextNonce
-	for _, txn := range txns {
-		if txn.Nonce < lowestNonce {
-			lowestNonce = txn.Nonce
-		}
-		t.Push(txn)
-	}
-	t.nextNonce = lowestNonce
-}
-
-// Add adds a new tx into the queue
-func (t *txQueue) Add(tx *types.Transaction) {
+// Add adds a new tx onto the account based tx heap
+func (t *txHeapWrapper) Add(tx *types.Transaction) {
 	t.Push(tx)
 }
 
-// Promote promotes all the new valid transactions
-func (t *txQueue) Promote() []*types.Transaction {
-	// Remove elements lower than nonce
+// pruneLowNonceTx removes any transactions from the account tx queue
+// that have a lower nonce than the current account nonce in state
+func (t *txHeapWrapper) pruneLowNonceTx() {
 	for {
+		// Grab the min-nonce transaction from the heap
 		tx := t.Peek()
 		if tx == nil || tx.Nonce >= t.nextNonce {
 			break
 		}
+
+		// Drop it from the heap
 		t.Pop()
 	}
+}
+
+// Promote promotes all the new valid transactions
+func (t *txHeapWrapper) Promote() []*types.Transaction {
+	// Remove elements lower than nonce
+	t.pruneLowNonceTx()
 
 	// Promote elements
 	tx := t.Peek()
 	if tx == nil || tx.Nonce != t.nextNonce {
+		// Nothing to promote
 		return nil
 	}
 
 	promote := []*types.Transaction{}
+	higherNonceTx := []*types.Transaction{}
+
+	reinsertFunc := func() {
+		// Reinsert the tx back to the account specific transaction queue
+		for _, highNonceTx := range higherNonceTx {
+			t.Push(highNonceTx)
+		}
+	}
+
 	for {
 		promote = append(promote, tx)
 		t.Pop()
 
-		tx2 := t.Peek()
-		if tx2 == nil || tx.Nonce+1 != tx2.Nonce {
+		nextTx := t.Peek()
+		if nextTx == nil {
+			break
+		}
```

Comment:
perhaps it's cleaner to include this in the `for` statement itself such as:

```
for nextTx := t.Peek(); nextTx != nil {
```

Reply:
I'm not sure what you mean by this `for` syntax. Did you mean `if`?

Reply:
Simplified 👍 
b812710

Reply:
@zivkovicmilos My bad, what I meant to suggest is:

```
for nextTx := t.Peek(); nextTx; nextTx != nil {
```

but it doesn't look as good 👍